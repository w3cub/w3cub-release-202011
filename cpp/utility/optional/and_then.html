
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Std&#58;&#58;optional&#58;&#58;and_then - C++ - W3cubDocs</title>
  
  <meta name="description" content="If &#42;this contains a value, invokes f with the contained value as an argument, and returns the result of that invocation; otherwise, returns an &hellip;">
  <meta name="keywords" content="std, optional, t, and, then, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/utility/optional/and_then.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/cpp.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _cppref">
				
				
<h1 id="firstHeading" class="firstHeading">std::optional&lt;T&gt;::and_then</h1>            <table class="t-dcl-begin"> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) &amp;;</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) const&amp;;</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) &amp;&amp;;</pre>
</td> <td> (3) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) const&amp;&amp;;</pre>
</td> <td> (4) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <p>If <code>*this</code> contains a value, invokes <code>f</code> with the contained value as an argument, and returns the result of that invocation; otherwise, returns an empty <code>std::optional</code>.</p>
<p>The return type (see below) must be a specialization of <code><a href="../optional" title="cpp/utility/optional">std::optional</a></code> (unlike <a href="transform" title="cpp/utility/optional/transform"><code>transform()</code></a>). Otherwise, the program is ill-formed.</p>
<div class="t-li1">
<span class="t-li">1)</span> Equivalent to<br><div class="cpp source-cpp"><pre data-language="cpp">if (*this)
    return std::invoke(std::forward&lt;F&gt;(f), this-&gt;value());
else
    return std::remove_cvref_t&lt;std::invoke_result_t&lt;F, T&amp;&gt;&gt;();</pre></div>
</div> <div class="t-li1">
<span class="t-li">2)</span> Equivalent to<br><div class="cpp source-cpp"><pre data-language="cpp">if (*this)
    return std::invoke(std::forward&lt;F&gt;(f), this-&gt;value());
else
    return std::remove_cvref_t&lt;std::invoke_result_t&lt;F, const T&amp;&gt;&gt;();</pre></div>
</div> <div class="t-li1">
<span class="t-li">3)</span> Equivalent to<br><div class="cpp source-cpp"><pre data-language="cpp">if (*this)
    return std::invoke(std::forward&lt;F&gt;(f), std::move(this-&gt;value()));
else
    return std::remove_cvref_t&lt;std::invoke_result_t&lt;F, T&gt;&gt;();</pre></div>
</div> <div class="t-li1">
<span class="t-li">4)</span> Equivalent to<br><div class="cpp source-cpp"><pre data-language="cpp">if (*this)
    return std::invoke(std::forward&lt;F&gt;(f), std::move(this-&gt;value()));
else
    return std::remove_cvref_t&lt;std::invoke_result_t&lt;F, const T&gt;&gt;();</pre></div>
</div>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> f </td> <td> - </td> <td> a suitable function or <a href="../../named_req/callable" title="cpp/named req/Callable">Callable</a> object that returns an <code><a href="../optional" title="cpp/utility/optional">std::optional</a></code> </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p>The result of <code>f</code> or an empty <code><a href="../optional" title="cpp/utility/optional">std::optional</a></code>, as described above.</p>
<h3 id="Notes"> Notes</h3> <p>Some languages call this operation <i>flatmap</i>.</p>
<table class="wikitable ftm-begin ftm-has-value"> <tr> <th>
<a href="../feature_test" title="cpp/utility/feature test">Feature-test</a> macro</th> <th>Value</th> <th><span title="Standard in which the feature is introduced">Std</span></th> <th>Comment </th>
</tr> <tr> <td><a href="../../feature_test#Library_features" title="cpp/feature test"><code>__cpp_lib_optional</code></a></td> <td><span class="nu0">202110L</span></td> <td><span class="t-mark">(C++23)</span></td> <td>
<a href="../optional#Monadic_operations" title="cpp/utility/optional">Monadic operations</a> in <code><a href="../optional" title="cpp/utility/optional">std::optional</a></code> </td>
</tr>
</table> <h3 id="Example"> Example</h3> <div class="t-example"> <div class="cpp source-cpp"><pre data-language="cpp">#include &lt;charconv&gt;
#include &lt;iomanip&gt;
#include &lt;iostream&gt;
#include &lt;optional&gt;
#include &lt;ranges&gt;
#include &lt;string&gt;
#include &lt;string_view&gt;
#include &lt;vector&gt;
 
std::optional&lt;int&gt; to_int(std::string_view sv)
{
    int r {};
    auto [ptr, ec] { std::from_chars(sv.data(), sv.data() + sv.size(), r) };
    if (ec == std::errc())
        return r;
    else
        return std::nullopt;
}
 
int main()
{
    using namespace std::literals;
 
    const std::vector&lt;std::optional&lt;std::string&gt;&gt; v
    {
        "1234", "15 foo", "bar", "42", "5000000000", " 5", std::nullopt, "-43"
    };
 
    for (auto&amp;&amp; x : v | std::views::transform(
        [](auto&amp;&amp; o)
        {
            // debug print the content of input optional&lt;string&gt;
            std::cout &lt;&lt; std::left &lt;&lt; std::setw(13)
                      &lt;&lt; std::quoted(o.value_or("nullopt")) &lt;&lt; " -&gt; ";
 
            return o
                // if optional is nullopt convert it to optional with "" string
                .or_else([]{ return std::optional{""s}; })
                // flatmap from strings to ints (making empty optionals where it fails)
                .and_then(to_int)
                // map int to int + 1
                .transform([](int n) { return n + 1; })
                // convert back to strings
                .transform([](int n) { return std::to_string(n); })
                // replace all empty optionals that were left by
                // and_then and ignored by transforms with "NaN"
                .value_or("NaN"s);
        }))
        std::cout &lt;&lt; x &lt;&lt; '\n';
}</pre></div> <p>Output:</p>
<div class="text source-text"><pre data-language="c">"1234"        -&gt; 1235
"15 foo"      -&gt; 16
"bar"         -&gt; NaN
"42"          -&gt; 43
"5000000000"  -&gt; NaN
" 5"          -&gt; NaN
"nullopt"     -&gt; NaN
"-43"         -&gt; -42</pre></div> </div> <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="value_or" title="cpp/utility/optional/value or"> <span class="t-lines"><span>value_or</span></span></a></div> </td> <td> returns the contained value if available, another value otherwise <br> <span class="t-mark">(public member function)</span>  </td>
</tr> <tr class="t-dsc"> <td> <div><a href="transform" title="cpp/utility/optional/transform"> <span class="t-lines"><span>transform</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx23">(C++23)</span></span></span></div> </td> <td> returns an <code>optional</code> containing the transformed contained value if it exists, or an empty <code>optional</code> otherwise <br> <span class="t-mark">(public member function)</span>  </td>
</tr> <tr class="t-dsc"> <td> <div><a href="or_else" title="cpp/utility/optional/or else"> <span class="t-lines"><span>or_else</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx23">(C++23)</span></span></span></div> </td> <td> returns the <code>optional</code> itself if it contains a value, or the result of the given function otherwise <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    © cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/optional/and_then" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/optional/and_then</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
