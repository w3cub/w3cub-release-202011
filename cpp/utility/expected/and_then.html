
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Std&#58;&#58;expected&#58;&#58;and_then - C++ - W3cubDocs</title>
  
  <meta name="description" content="If &#42;this contains an expected value, invokes f and returns its result; otherwise, returns a std&#58;&#58;expected object that contains a copy of &hellip;">
  <meta name="keywords" content="std, expected, t, e, and, then, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/utility/expected/and_then.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/cpp.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _cppref">
				
				
<h1 id="firstHeading" class="firstHeading">std::expected&lt;T,E&gt;::and_then</h1>            <table class="t-dcl-begin"> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) &amp;;</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) const&amp;;</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) &amp;&amp;;</pre>
</td> <td> (3) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class F &gt;
constexpr auto and_then( F&amp;&amp; f ) const&amp;&amp;;</pre>
</td> <td> (4) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <p>If <code>*this</code> contains an expected value, invokes <code>f</code> and returns its result; otherwise, returns a <code>std::expected</code> object that contains a copy of <a href="error" title="cpp/utility/expected/error"><code>error()</code></a>.</p>
<p>If <code>T</code> is not (possibly cv-qualified) <span class="kw4">void</span>, the contained value (<a href="value" title="cpp/utility/expected/value"><code>value()</code></a>) is passed as an argument to <code>f</code>; otherwise <code>f</code> takes no argument.</p>
<p>Let <code>U</code> be:</p>
<ul>
<li> if <code>T</code> is not (possibly cv-qualified) <span class="kw4">void</span>: <ul>
<li> for overloads <span class="t-v">(1-2)</span>, <code><a href="http://en.cppreference.com/w/cpp/types/remove_cvref"><span class="kw2648">std::<span class="me2">remove_cvref_t</span></span></a><span class="sy1">&lt;</span><a href="http://en.cppreference.com/w/cpp/types/result_of"><span class="kw3223">std::<span class="me2">invoke_result_t</span></span></a><span class="sy1">&lt;</span>F, decltype<span class="br0">(</span>value<span class="br0">(</span><span class="br0">)</span><span class="br0">)</span><span class="sy1">&gt;&gt;</span></code>; </li>
<li> for overloads <span class="t-v">(3-4)</span>, <code><a href="http://en.cppreference.com/w/cpp/types/remove_cvref"><span class="kw2648">std::<span class="me2">remove_cvref_t</span></span></a><span class="sy1">&lt;</span><a href="http://en.cppreference.com/w/cpp/types/result_of"><span class="kw3223">std::<span class="me2">invoke_result_t</span></span></a><span class="sy1">&lt;</span>F, decltype<span class="br0">(</span>std<span class="sy4">::</span><span class="me2">move</span><span class="br0">(</span>value<span class="br0">(</span><span class="br0">)</span><span class="br0">)</span><span class="br0">)</span><span class="sy1">&gt;&gt;</span></code>; </li>
</ul> </li>
<li> otherwise (<code>T</code> is possibly cv-qualified <span class="kw4">void</span>), <code><a href="http://en.cppreference.com/w/cpp/types/remove_cvref"><span class="kw2648">std::<span class="me2">remove_cvref_t</span></span></a><span class="sy1">&lt;</span><a href="http://en.cppreference.com/w/cpp/types/result_of"><span class="kw3223">std::<span class="me2">invoke_result_t</span></span></a><span class="sy1">&lt;</span>F<span class="sy1">&gt;&gt;</span></code>. </li>
</ul> <p>The return type is <code>U</code>, which must be a specialization of <code>std::expected</code>, and <code><a href="http://en.cppreference.com/w/cpp/types/is_same"><span class="kw592">std::<span class="me2">is_same_v</span></span></a><span class="sy1">&lt;</span>U<span class="sy4">::</span><span class="me2">error_type</span>, E<span class="sy1">&gt;</span></code> must be <span class="kw2">true</span>.</p>
<pre data-language="cpp">true</pre> <pre data-language="cpp">true</pre>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> f </td> <td> - </td> <td> a suitable function or <a href="../../named_req/callable" title="cpp/named req/Callable">Callable</a> object that returns a <code>std::expected</code> </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p>The result of <code>f</code> or a <code>std::expected</code> object that contains an error value, as described above.</p>
<h3 id="Notes"> Notes</h3> <table class="wikitable ftm-begin ftm-has-value"> <tr> <th>
<a href="../feature_test" title="cpp/utility/feature test">Feature-test</a> macro</th> <th>Value</th> <th><span title="Standard in which the feature is introduced">Std</span></th> <th>Comment </th>
</tr> <tr> <td><a href="../../feature_test#Library_features" title="cpp/feature test"><code>__cpp_lib_expected</code></a></td> <td><span class="nu0">202211L</span></td> <td><span class="t-mark">(C++23)</span></td> <td>Monadic functions for <code>std::expected</code> </td>
</tr>
</table> <h3 id="Example"> Example</h3>  <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="unexpect_t" title="cpp/utility/expected/unexpect t"> <span class="t-lines"><span>unexpect_t</span><span>unexpect</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx23">(C++23)</span></span></span></div> </td> <td> in-place construction tag for unexpected value in <code>expected</code> <br> <span class="t-mark">(class)</span> <span class="t-mark">(constant)</span>  </td>
</tr> <tr class="t-dsc"> <td> <div><a href="transform" title="cpp/utility/expected/transform"> <span class="t-lines"><span>transform</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx23">(C++23)</span></span></span></div> </td> <td> returns an <code>expected</code> containing the transformed expected value if it exists; otherwise, returns the <code>expected</code> itself <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    Â© cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/expected/and_then" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/expected/and_then</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
